src/finn/transformation/change_3d_tensors_to_4d.py

```diff
@@ -0,0 +1,219 @@
+        if len(invalid_nodes) > 0:
+            warnings.warn(
+                "Transformation is not applied,\
+                 found unsupported nodes in the graph: {}.".format(
+                    invalid_nodes
+                )
+            )
+            return (model, graph_modified)
+
+        # Infer the shapes of each tensor, remove unused tensors
+        # and give each tensor a readable name
+        model = model.transform(InferShapes())
+        model = model.transform(RemoveUnusedTensors())
+        model = model.transform(GiveUniqueNodeNames())
+        model = model.transform(GiveReadableTensorNames())
+
+        # Converts 3D tensors representing the input, a value_info, or the output
+        # to 4D tensors
+        # Inputs
+        tensor_names = {}
+        for t in model.graph.input:
+            tensor_name = t.name
+            tensor_type = t.type.tensor_type.elem_type
+            tensor_shape = model.get_tensor_shape(tensor_name)
+            tensor_names[tensor_name] = [tensor_type]
+            tensor_names[tensor_name].append(tensor_shape)
+
+        # Initializers
+        initializer_names = []  # list with all initializers
+        initializers = {}  # list with initializers from Conv, Mul, and Add nodes
+        for i in model.graph.initializer:
+            init_name = i.name
+            initializer_names.append(init_name)
+            if "Conv" in init_name:
+                init_dim = i.dims
+                init_dtype = i.data_type
+                initializers[init_name] = [init_dtype]
+                initializers[init_name].append(init_dim)
+            elif init_name[0:4] == "Mul_":
+                init_dim = i.dims
+                if len(i.dims) == 3:
+                    init_dtype = i.data_type
+                    initializers[init_name] = [init_dtype]
+                    initializers[init_name].append(init_dim)
+            elif "Add" in init_name:
+                init_dim = i.dims
+                if len(i.dims) == 3:
+                    init_dtype = i.data_type
+                    initializers[init_name] = [init_dtype]
+                    initializers[init_name].append(init_dim)
+
+        # Value infos
+        for t in model.graph.value_info:
+            tensor_name = t.name
+            if tensor_name in initializer_names:
+                continue
+            else:
+                tensor_type = t.type.tensor_type.elem_type
+                tensor_shape = model.get_tensor_shape(tensor_name)
+                tensor_names[tensor_name] = [tensor_type]
+                tensor_names[tensor_name].append(tensor_shape)
+
+        # Outputs
+        for t in model.graph.output:
+            tensor_name = t.name
+            tensor_type = t.type.tensor_type.elem_type
+            tensor_shape = model.get_tensor_shape(tensor_name)
+            tensor_names[tensor_name] = [tensor_type]
+            tensor_names[tensor_name].append(tensor_shape)
```

Comment:
This bit (plus the bit that extracts the initializers above) could potentially be simplified by iterating on the list returned from `ModelWrapper.get_all_tensor_names()` -- a list of all input, output and value_info tensor names in the graph. You can then iterate over this list to check whether it needs to be considered for 3D->4D or not.